wandb/sdk/artifacts/storage_policies/wandb_storage_policy.py

```
@@ -165,6 +212,129 @@ def load_file(
                 file.write(data)
         return path
 
+    def _should_multipart_download(
+        self,
+        file_size: int | None,
+        multipart: bool | None,
+    ) -> bool:
+        if file_size is None:
+            return False
+        if multipart is not None:
+            return multipart
+        return file_size >= _MULTIPART_DOWNLOAD_SIZE
+
+    def _write_chunks_to_file(
+        self,
+        f: IO,
+        q: queue.Queue,
+        download_has_error: threading.Event,
+    ):
+        while not download_has_error.is_set():
+            item = q.get()
+            if item is _CUHNK_QUEUE_SENTIENL:
+                # Normal shutdown, all the chunks are written
+                return
+            elif isinstance(item, _ChunkContent):
+                try:
+                    # NOTE: Seek works without pre allocating the file on disk.
+                    # It automatically creates a sparse file, e.g. ls -hl would show
+                    # a bigger size compared to du -sh * because downloading different
+                    # chunks is not a sequential write.
+                    # See https://man7.org/linux/man-pages/man2/lseek.2.html
+                    f.seek(item.offset)
+                    f.write(item.data)
+                except Exception as e:
+                    download_has_error.set()
+                    raise e
+            else:
+                raise ValueError(f"Unknown queue item type: {type(item)}")
+
+    def _download_part(
+        self,
+        download_url: str,
+        headers: dict,
+        start: int,
+        q: queue.Queue,
+        download_has_error: threading.Event,
+    ):
+        # Other threads has error, no need to start
+        if download_has_error.is_set():
+            return
+        response = self._session.get(
+            url=download_url,
+            headers=headers,
+            stream=True,
+        )
+        response.raise_for_status()
```

question: are the errors from this response helpful for the user? or do we need to raise a better error?

---

I am coping existing serial download logic here, I think the status code error is human readable https://3.python-requests.org/user/quickstart/#response-status-codes

Though one difference with current logic is I am not falling back to bear token authorization header when signed url fails.
I am not sure if it worth to add the fallback, e.g. if multipart download has http status error or fail for whatever reason, fallback to http serial/http with bear access token.

> requests.exceptions.HTTPError: 404 Client Error

https://github.com/wandb/wandb/blob/af632002c86a70f7f1a3510a8b8517f5179fc2dd/wandb/sdk/artifacts/storage_policies/wandb_storage_policy.py#L134-L166

---

not blocking: i know we have a ticket for better erroring in SDK this might fall under that